src/master.py

```diff
@@ -141,12 +103,109 @@ def work():
     bootstrap.update_ec2_stack(stackname, context, concurrency='serial')
     return True
 
-# TODO: extract just the salt update part from `remaster`
 @debugtask
 @requires_aws_stack
 def update_salt(stackname):
-    current_master_stack = core.find_master_for_stack(stackname)
-    return remaster(stackname, current_master_stack)
+    "updates the installed version of salt to the version specified in the project file. see `docs/salt.md`"
+    # TODO: turn this into a decorator
+    import cfn
+    # start the machine if it's stopped
+    # you might also want to acquire a lock so alfred doesn't stop things or update things
+    cfn._check_want_to_be_running(stackname, 1)
+
+    # complete knowledge of a stack's particulars
+    context = context_handler.load_context(stackname)
+    if not context.get('ec2'):
+        LOG.info("no ec2 context: %s", stackname)
+        return
+
+    # 1. get the value of the desired version of salt from the project config file (pdata)
+    # 2. update the stack's current context and write that back to s3
+    # 3. update the subset of the context on the ec2 instance itself (buildvars)
+    # 4. run highstate
+    LOG.info("updating salt version in context file")
+    pdata = core.project_data_for_stackname(stackname)
+    context['project']['salt'] = pdata['salt']
+    context_handler.write_context(stackname, context)
+
+    LOG.info("updating buildvars on ec2 instance")
+    buildvars.refresh(stackname, context)
```

Comment:
`buildvars` are essentially a full copy of the `context`; if there is an update of infrastructure pending, we are propagating the new `context` to the instance without necessarily having created/updated/destroyed that infrastructure. Since [the formulas extensively refer to `buildvars`](https://github.com/elifesciences/builder-base-formula/search?q=elife.cfg&unscoped_q=elife.cfg) to decide what to do, I see this optimization as unsafe.

As an alternative we could remove the Salt version from the `context` to make it more dynamic, but that isn't right either because we wanted a controlled upgrade.

Last alternative is to use `update_infrastructure` and make sure we avoid drift. Should we calculate drifts in Alfred and publish some kind of warning when there are some?

Reply:
this is good feedback but unfortunately part of this changeset made it into master. 

I've opened a [branch](https://github.com/elifesciences/builder/pull/665) to deal with this for the next upgrade.